### PR TITLE
Ledger: Limit TX set size in nominations

### LINKED ISSUE #2712: Block size should be limited
It will probably take some time reach to a level of network activity for this to matter

### PR DESCRIPTION
**Pull request by omerfirmak**
Fixes #2712

Point made in https://github.com/bosagora/agora/pull/2416#issuecomment-906494622 still stands, will need a separate PR

### CODE REVIEW

**Comment by omerfirmak** [comment]
Added a second commit to address https://github.com/bosagora/agora/pull/2416#issuecomment-906494622

File: source/agora/consensus/Ledger.d [hunk 1]
```diff
@@ -228,7 +228,8 @@ public class NodeLedger : Ledger
         if (auto err = this.fee_man.getTxFeeRate(tx, this.utxo_set.getUTXOFinder(),
             &this.getPenaltyDeposit, fee_rate))
             return err;
-        if (fee_rate < min_fee)
+        const size_budget = this.params.MaxTxSetSize * 1024;
```
Comment (hewison-chris) [inline source/agora/consensus/Ledger.d]:
```suggestion
        const size_limit = this.params.MaxTxSetSize * 1024;
```

File: source/agora/consensus/Ledger.d [hunk 2]
```diff
@@ -228,7 +228,8 @@ public class NodeLedger : Ledger
...
+        if (fee_rate < min_fee && this.pool.getPoolSize() >= size_budget) // If pool is sparsely populated, accept any fee.
```
Comment (hewison-chris) [inline source/agora/consensus/Ledger.d]: Isn't there a risk that no fees will be paid with this change?

File: source/agora/consensus/Ledger.d [hunk 3]
```diff
@@ -365,6 +366,11 @@ public class NodeLedger : Ledger
         if (auto fail_reason = this.getValidTXSet(data, tx_set, utxo_finder))
             return fail_reason;
 
+        const tx_set_size = tx_set.map!(tx => tx.sizeInBytes()).sum();
+        const size_budget = this.params.MaxTxSetSize * 1024;
```
Comment (hewison-chris) [inline source/agora/consensus/Ledger.d]:
```suggestion
        const size_limit = this.params.MaxTxSetSize * 1024;
```

File: source/agora/consensus/Ledger.d [hunk 4]
```diff
@@ -757,11 +763,11 @@ public class ValidatingLedger : NodeLedger
 
     ***************************************************************************/
 
-    public Hash[] getCandidateTransactions (in Height height, ulong max_txs,
-        scope UTXOFinder utxo_finder) @safe
+    public Hash[] getCandidateTransactions (in Height height, scope UTXOFinder utxo_finder) @safe
     {
         Hash[] result;
         Amount tot_fee, tot_data_fee;
+        size_t size_budget = this.params.MaxTxSetSize * 1024; // cant overflow
```
Comment (hewison-chris) [inline source/agora/consensus/Ledger.d:784]:
```suggestion
        size_t size_limit = this.params.MaxTxSetSize * 1024; // cant overflow
```

File: source/agora/consensus/Ledger.d [hunk 5]
```diff
@@ -861,33 +867,28 @@ public class ValidatingLedger : NodeLedger
     }
 
     /// Generate a new block by creating transactions, then calling `forceCreateBlock`
-    private Transaction[] makeTestBlock (
-        Transaction[] last_txs, ulong txs = Block.TxsInTestBlock)
+    private Transaction[] makeTestBlock (Transaction[] last_txs)
```
Comment (hewison-chris) [inline source/agora/consensus/Ledger.d:884]: Nice, finally removed. :-)

**Review by hewison-chris — APPROVED** [review]
LGTM, just not sure we should allow lower then min fee rate when tx set size is below the limit.

File: source/agora/consensus/Ledger.d [hunk 6]
```diff
@@ -757,11 +762,11 @@ public class ValidatingLedger : NodeLedger
 
     ***************************************************************************/
 
-    public Hash[] getCandidateTransactions (in Height height, ulong max_txs,
-        scope UTXOFinder utxo_finder) @safe
+    public Hash[] getCandidateTransactions (in Height height, scope UTXOFinder utxo_finder) @safe
```
Comment (linked0) [inline source/agora/consensus/Ledger.d]: Two-line is better, I guess.

File: source/agora/consensus/Ledger.d [hunk 7]
```diff
@@ -1594,6 +1596,29 @@ unittest
     auto average_tx = genesisSpendable().front().refund(WK.Keys[0].address).deduct(10.coins).sign();
     assert(ledger.acceptTransaction(average_tx, 0, min_fee_pct) is null);
 
+    // switch to a different input, even with low fees this should be accepted since pool is almost empty
```
Comment (linked0) [inline source/agora/consensus/Ledger.d]: All the added lines in the bracket should be indented.

File: source/agora/consensus/pool/Transaction.d [hunk 8]
```diff
@@ -475,6 +475,25 @@ public class TransactionPool
         return 0.coins;
     }
 
+    /***************************************************************************
+
+        Returns:
+            Total size of the transactions in the pool
```
Comment (linked0) [inline source/agora/consensus/pool/Transaction.d]: Would be better to add the unit of size, like "in bytes"

File: source/agora/consensus/Ledger.d [hunk 9]
```diff
@@ -1727,6 +1727,51 @@ unittest
     }
 
     // should be able to create a new block after previous height receives majority signatures
-    ledger.forceCreateBlock(0);
+    ledger.forceCreateBlock();
     assert(ledger.getBlockHeight() == 2);
 }
+
+unittest
```
Comment (linked0) [inline source/agora/consensus/Ledger.d]: Isn't it better to explain what this unit test is for?

**Review by linked0 — COMMENTED** [review]
Broadly good. Left some suggestions.
And please add the commit message on the first commit.

File: source/agora/consensus/Ledger.d (hunk 2 above)
Comment (omerfirmak) [reply source/agora/consensus/Ledger.d]: I will add a test for it.

Comment (omerfirmak) [reply source/agora/consensus/Ledger.d]: added a simple check for it, FeeManager handles that and does not allow any TXs with fees below `ConsensusConfig.min_fee`

**Comment by omerfirmak** [comment]
> not sure we should allow lower then min fee rate when tx set size is below the limit

We need that to be able to talk about an average of fees in the pool. If we don't do that, a single high paying TX can block any other TX to be accepted in a block.

File: source/agora/consensus/Ledger.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/consensus/Ledger.d:784]: I applied previous suggestions but since we keep subtracting from this, it actually represents the budget we have left. So it seem like a better name.

File: source/agora/consensus/Ledger.d (hunk 7 above)
Comment (omerfirmak) [reply source/agora/consensus/Ledger.d]: They result in a hard to follow diff tho

Comment (linked0) [reply source/agora/consensus/Ledger.d]: Indentation is the basic requirement, I guess. But it's not the critical one.

**Review by linked0 — APPROVED** [review]
LGTM

**Comment by omerfirmak** [comment]
Rebased
